Patches for OSX compilation warnings

mheiber opened the pull request:
Fix compilation warnings caught on OS X:
- NormalizeFunc was defined twice: remove the duplicate and fix
up `#include`s
- Avoid unsigned integer overflow bug vector in trie_type.c:
`slen-len` isn't safe when the lengths are unsigned.

Also modified the main Makefile to use the `-Werror` flag in all environments, since OS X doesn't complain anymore.

mheiber commented on src/trie/trie_type.c:
```diff
@@ -88,7 +88,7 @@ Vector *Trie_Search(Trie *tree, char *s, size_t len, size_t num, int maxDist, in
     }
     // in prefix mode we also factor in the total length of the suffix
     if (prefixMode) {
-      ent->score /= sqrt(1 + fabs(slen - len));
```
If I remember correctly, clang threw a warning for `fabs` here because `slen` and `len` are unsigned. But getting rid of `fabs` isn't the right solution, I should have cast each of `slen` and `len` to a signed type instead and kept the call to `fabs`. Nice catch. Will fix.

mheiber replied:
Actually, it's not safe to cast size_t to an unsigned type, so now I'm just subtracting the lesser number from the greater number.

dvirsky replied:
size_t is an unsigned type though.

mheiber replied:
Whoops, I meant it's not safe to cast size_t to a signed type, since the standard doesn't specify what that type would be. And subtraction between unsigned types could underflow.

dvirsky replied:
ok, what you did makes sense

mheiber commented:
@dvirsky I'm getting this test failure, are you seeing this too?

    test_range.c:17: 46 != 60:

dvirsky commented:
@mheiber re the test failure - nope, the test failure is not happening for me. can you make sure you make clean before testing this?

dvirsky commented on src/Makefile:
```diff
@@ -13,6 +13,7 @@ else
 	SHOBJ_CFLAGS ?= -dynamic -fno-common -g -ggdb
 	SHOBJ_LDFLAGS ?= -bundle -undefined dynamic_lookup
 endif
+
```
lines 10+11
```c
# We add -Werror to CFLAGS only on Linux since on OSX there are still tons of warnings
CFLAGS +=-Werror  
```
for linux only  should be removed if -Werror is now global

dvirsky commented:
@mheiber I've managed to get my hands on a Mac, and tried compiling it, got tons of linker errors, aren't you getting those?

In any case, I've googled a bit and found the fix, now I just need to be able to commit from that machine :)

mheiber commented:
I got tons of linker errors as well. What worked for me was getting rid of
the -lc -lm . What worked for you?

On Tue, Jan 31, 2017 at 3:01 PM, Dvir Volk <notifications@github.com> wrote:

> @mheiber <https://github.com/mheiber> I've managed to get my hands on a
> Mac, and tried compiling it, got tons of linker errors, aren't you getting
> those?
>
> In any case, I've googled a bit and found the fix, now I just need to be
> able to commit from that machine :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RedisLabsModules/RediSearch/pull/24#issuecomment-276475248>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEeUlSvXmBxoQ5Hu7Zo_1xNvF2BaBWwXks5rX5M1gaJpZM4Lqfif>
> .
>

dvirsky commented:
Nope. I removed them from CFLAGS, indeed they shouldn't be there. They should be in the linker flags. But what I got was tons of errors when linking against Snowball stemmer, about incompatible mac os versions. I managed to address this issue specifically and got rid of them.

mheiber commented:
Was it this issue? "was built for newer OSX version (10.12) than being
linked (10.10)"
I fixed that by setting this environment variable on my machine: `
MACOSX_DEPLOYMENT_TARGET=10.10`, but still don't understand how the OSX
targets get out of sync.

On Tue, Jan 31, 2017 at 4:15 PM, Dvir Volk <notifications@github.com> wrote:

> Nope. I removed them from CFLAGS, indeed they shouldn't be there. They
> should be in the linker flags. But what I got was tons of errors when
> linking against Snowball stemmer, about incompatible mac os versions. I
> managed to address this issue specifically and got rid of them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RedisLabsModules/RediSearch/pull/24#issuecomment-276494835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEeUlYYL2PZTq8tbaZLko9D9E4WGbimuks5rX6SHgaJpZM4Lqfif>
> .
>

dvirsky commented:
@mheiber yep.
The fix was to add `-mmacosx-version-min=10.6` to the CFLAGS in `dep/snowball/GNUMakeFile`, and `-mmacosx_version_min 10.6` to the SHOBJ_LDFLAGS in the root Makefile (notice the two forms of writing it, ld accepts different flags than gcc), only in the mac os part.

mheiber commented:
@dvirsky this PR currently includes:
- changes to flags
- changes to code

Since you refactored flags and are testing on OS X, I plan to modify this PR to include only code changes. 👍 ?

dvirsky commented:
I'd be happy with both, but maybe just rebase first because I made a lot of changes to compilation flags. As long as no warnings are issued and tests all pass, I'm good. 

Notice that right now I'm exporting CFLAGS in the main make file, and on all sub makefiles, I've changed to `CFLAGS ?= ...` which means that when compiled from the main makefile, we simply use the root makefile's CFLAGS.

mheiber commented:
I'm still getting `test_range.c:17: 46 != 60: ` on OSX, even on a fresh pull. I'm not seeing the problem on Ubuntu. I'll dig in.

dvirsky commented:
@mheiber I should be getting a new Mac this week (yay!). you can leave it and I'll try to see what's up. it's interesting though that we have a bug and not just a compiler issue.

mheiber commented:
@dvirsky the problem may be with line https://github.com/RedisLabsModules/RediSearch/blob/master/src/tests/test_range.c#L15. `rand()` returned different values in OS X as compared to Ubuntu, though it was deterministic in both environments.

dvirsky commented:
@mheiber I've seeded `srand()` to get consistent results. We can substitute this with a custom pseudo random function, but since it's just a problem with the test, let's ignore this failure for now, and I'll handle it when I switch to Mac. 

So what's the status of the PR? Can you rebase and squash and let's finish with this?

mheiber commented:
Sounds good, I can wrap it up tomorrow.

On Mon, Feb 6, 2017 at 2:22 PM, Dvir Volk <notifications@github.com> wrote:

> @mheiber <https://github.com/mheiber> I've seeded srand() to get
> consistent results. We can substitute this with a custom pseudo random
> function, but since it's just a problem with the test, let's ignore this
> failure for now, and I'll handle it when I switch to Mac.
>
> So what's the status of the PR? Can you rebase and squash and let's finish
> with this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RedisLabsModules/RediSearch/pull/24#issuecomment-277784348>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEeUlSAf-KL5i45MvQVq4hveg46PBtBTks5rZ3LngaJpZM4Lqfif>
> .
>

dvirsky commented:
@mheiber I've added a custom PRNG function (Blum Blum Shub) to the tests (not pushed yet), if you want just replace it in the test.

```diff
+u_int prng_seed = 1337;
+#define PRNG_MOD 30980347
+u_int prng() {
+  prng_seed = (prng_seed * prng_seed) % PRNG_MOD;
+  return prng_seed;
+}
+
 }

+u_int prng_seed = 1337;
+#define PRNG_MOD 30980347
+u_int prng() {
+  prng_seed = (prng_seed * prng_seed) % PRNG_MOD;
+  return prng_seed;
+}
+
 int testRangeIterator() {
   NumericRangeTree *t = NewNumericRangeTree();
   ASSERT(t != NULL);

-  srand(1337);
   int N = 100;
   for (int i = 0; i < N; i++) {

     t_docId docId = i + 1;
-    float value = (double)(1 + rand() % 1000);
+    float value = (double)(1 + prng() % 1000);
     // printf("Adding %d > %f\n", docId, value);
     NumericRangeTree_Add(t, docId, value);
   }
-  // ASSERT_EQUAL(t->numRanges, 13);
+  ASSERT_EQUAL(t->numRanges, 8);
   ASSERT_EQUAL(t->numEntries, N);

   NumericFilter *flt = NewNumericFilter(-1, 1002, 0, 0);
```

mheiber commented:
@dvirsky I fixed up the compiler errors.
Re the test: could we use `srand` to set the random seed?

dvirsky commented:
I used srand in the previous version, it looks like the rand() function itself might work differently on OSX. Anyway the custom function I made is pre seeded with a fixed number so will always give the same sequence of random numbers.

mheiber commented:
@dvirsky the `prng()` function did the trick, tests are passing for me on both OS X and Ubuntu now.

dvirsky commented:
Awesome. so are we good? Can you rebase and squash the entire branch? IIRC I've pushed the prng to master yesterday, so you'll probably see a collision.

mheiber commented:
We're good, I rebased again and squashed.
This new test failure seems to be unrelated:

```

Assertion failed: (strcmp(t.s, tx->expected[tx->num++]) == 0), function
tokenFunc, file test_index.c, line 377.

```

I'm seeing the same thing on `master` with Linux and on `master` with OS X.

On Wed, Feb 8, 2017 at 10:47 AM, Dvir Volk <notifications@github.com> wrote:

> Awesome. so are we good? Can you rebase and squash the entire branch? IIRC
> I've pushed the prng to master yesterday, so you'll probably see a
> collision.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RedisLabsModules/RediSearch/pull/24#issuecomment-278365810>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEeUlbQ5LzjrRCDPcrMgu-5CK34Dro_Eks5raeOMgaJpZM4Lqfif>
> .
>

dvirsky commented:
@mheiber thanks, I fixed the tests. I'll pull in a bit.

dvirsky commented:
Merged. Thanks a lot!

mheiber commented:
@dvirsky thanks for the coaching

dvirsky commented:
@mheiber I'm not seeing the mac os CFLAGS. did you add those?

mheiber commented:
@dvirsky I moved the `Werror` to the CFLAGS that apply in all environments:

https://github.com/RedisLabsModules/RediSearch/pull/24/files#diff-f57f2991a6aa25fe45d8036c51bf8b4dL5

dvirsky commented:
There was the issue with the -mmacosx-version-min flag that cause all the linker errors. nevermind, I'll patch it up myself
